### PR TITLE
TensorListView generalized reshape and reinterpret

### PR DESCRIPTION
**Pull request by mzient**
Add TensorListView reinterpret and reshape with sample split/merge even for non-contiguous lists.

Signed-off-by: Michał Zientkiewicz <mzient@gmail.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature needed to split sequences into separate frames for VideoResize

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     * `reshape` is now a special case of `reinterpret` with output type same as input type
     * `reinterpret` advances a pointer and splits/merges samples as necessary; error is raised only if the input discontinuity
       appears inside an output sample
 - Affected modules and functionalities:
     * reshape function and its users (spectrogram, reductions)
 - Key points relevant for the review:
     * Try to find some corner cases not handled by the algorithm?
 - Validation and testing:
     * GTest (tests include error cases)
 - Documentation (including examples):
     * Doxygen

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: include/dali/core/tensor_view.h [hunk 1]
```diff
@@ -778,43 +778,95 @@ TensorListView<StorageBackend, DataType, output_ndim> sample_range(
...
-                               "non-contiguous TensorListView");
-      assert(volume(list.tensor_shape_span(i)) == volume(shape.tensor_shape_span(i)));
+template <typename U, int out_dim, typename Storage, typename T, int in_dim>
+TensorListView<Storage, U, out_dim> reinterpret(
+      const TensorListView<Storage, T, in_dim> &list,
+      TensorListShape<out_dim> shape,
+      bool check = false) {
+  if (check && shape.num_elements() * sizeof(U) != list.num_elements() * sizeof(T))
+    throw std::logic_error("Attempt to reshape a TensorListView to a different total size");
+  assert(shape.num_elements() * sizeof(U) == list.num_elements() * sizeof(T));
```
Comment (jantonguirao) [inline include/dali/core/tensor_view.h:802]: why do you check this again?

File: include/dali/core/tensor_view.h [hunk 2]
```diff
@@ -778,43 +778,95 @@ TensorListView<StorageBackend, DataType, output_ndim> sample_range(
...
+
+  ptrdiff_t in_remaining = 0;
+  uintptr_t ptr = reinterpret_cast<uintptr_t>(list.data[0]);
+
+  TensorListView<Storage, U, out_dim> out_list;
+  out_list.shape = std::move(shape);
+  out_list.data.resize(N);
+
+  for (int o = 0; o < N; o++) {
+    ptrdiff_t out_remaining = volume(out_list.shape.tensor_shape_span(o)) * sizeof(U);
```
Comment (jantonguirao) [inline include/dali/core/tensor_view.h:819]: I added a `tensor_size()` method that does this exactly

File: include/dali/core/tensor_view.h [hunk 3]
```diff
@@ -778,43 +778,95 @@ TensorListView<StorageBackend, DataType, output_ndim> sample_range(
...
+    out_list.data[o] = reinterpret_cast<U *>(ptr);
+    bool first_chunk_in_sample = true;
+    while (out_remaining > 0) {
+      if (in_remaining > 0) {
+        ptrdiff_t to_add = out_remaining < in_remaining ? out_remaining : in_remaining;
```
Comment (jantonguirao) [inline include/dali/core/tensor_view.h:824]:
```suggestion
        ptrdiff_t to_add =std::min(out_remaining, in_remaining);
```

File: include/dali/core/tensor_view.h (hunk 2 above)
Comment (JanuszL) [reply include/dali/core/tensor_view.h:819]: Has it been merged already?

File: include/dali/core/tensor_view.h (hunk 3 above)
Comment (mzient) [reply include/dali/core/tensor_view.h:824]: That was (sort-of) deliberate `<algorithm>` is not included here and min/max are unfortunately there and not in `<utility>` :(
